File: backend/app/domain/templates.py
"""Prompt templates: ``{{ path }}`` placeholders resolved against a test case input.

Deliberately tiny (no logic, no filters): prompt templates should be readable at a glance and
render identically forever, which matters for reproducing historical runs.

* ``{{ question }}`` — a field of an object input (dotted paths and ``[0]`` indices work)
* ``{{ input }}`` — the whole input (strings as-is, structured values as pretty JSON)
"""

import json
import re
from typing import Any

PLACEHOLDER_RE = re.compile(r"\{\{\s*([A-Za-z_][\w.\[\]]*)\s*\}\}")
_PATH_TOKEN_RE = re.compile(r"([^.\[\]]+)|\[(\d+)\]")


class TemplateError(ValueError):
    pass


def template_variables(template: str) -> list[str]:
    return list(dict.fromkeys(PLACEHOLDER_RE.findall(template)))


def render_template(template: str, value: Any) -> str:
    def substitute(match: re.Match[str]) -> str:
        path = match.group(1)
        resolved = value if path == "input" else _resolve(value, path)
        return resolved if isinstance(resolved, str) else json.dumps(resolved, indent=2)

    return PLACEHOLDER_RE.sub(substitute, template)


def _resolve(value: Any, path: str) -> Any:
    current = value
    for match in _PATH_TOKEN_RE.finditer(path):
        key, index = match.groups()
        if key is not None and isinstance(current, dict) and key in current:
            current = current[key]
        elif index is not None and isinstance(current, list) and int(index) < len(current):
            current = current[int(index)]
        else:
            raise TemplateError(f"Template variable '{{{{ {path} }}}}' is missing from the input")
    return current

File: frontend/src/demo/engine/templates.ts
/** Port of `app/domain/templates.py`: `{{ path }}` placeholders resolved against a case input. */

import { isDict } from "./py";
import { dumps } from "./pyjson";

const PLACEHOLDER_RE = /\{\{\s*([A-Za-z_][\p{L}\p{N}_.[\]]*)\s*\}\}/gu;
const PATH_TOKEN_RE = /([^.[\]]+)|\[(\d+)\]/g;

export class TemplateError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "TemplateError";
  }
}

export function templateVariables(template: string): string[] {
  return [...new Set(Array.from(template.matchAll(PLACEHOLDER_RE), (m) => m[1]))];
}

export function renderTemplate(template: string, value: unknown): string {
  return template.replace(PLACEHOLDER_RE, (_match, path: string) => {
    const resolved = path === "input" ? value : resolve(value, path);
    return typeof resolved === "string" ? resolved : dumps(resolved, { indent: 2 });
  });
}

function resolve(value: unknown, path: string): unknown {
  let current = value;
  for (const [, key, index] of path.matchAll(PATH_TOKEN_RE)) {
    if (key !== undefined && isDict(current) && Object.hasOwn(current, key)) {
      current = current[key];
    } else if (index !== undefined && Array.isArray(current) && Number(index) < current.length) {
      current = current[Number(index)];
    } else {
      throw new TemplateError(`Template variable '{{ ${path} }}' is missing from the input`);
    }
  }
  return current;
}
